zeroize/derive/src/lib.rs

```diff
@@ -134,6 +160,49 @@ impl ZeroizeAttrs {
             };
 
             self.drop = true;
+        } else if meta.path().is_ident("bound") {
+            assert!(self.bound.is_none(), "duplicate #[zeroize] bound flags");
+
+            match (variant, binding) {
+                (_variant, Some(_binding)) => {
+                    // structs don't have a variant prefix, and only structs have bindings outside of a variant
+                    let item_kind = match variant.and_then(|variant| variant.prefix) {
+                        Some(_) => "enum",
+                        None => "struct",
+                    };
+                    panic!(
+                        concat!(
+                            "The #[zeroize(bound)] attribute is not allowed on {} fields. ",
+                            "Use it on the containing {} instead.",
+                        ),
+                        item_kind, item_kind,
+                    )
+                }
+                (Some(_variant), None) => panic!(concat!(
+                    "The #[zeroize(bound)] attribute is not allowed on enum variants. ",
+                    "Use it on the containing enum instead.",
+                )),
+                (None, None) => {
+                    if let Meta::NameValue(meta_name_value) = meta {
+                        if let Lit::Str(lit) = &meta_name_value.lit {
+                            if lit.value().is_empty() {
+                                self.bound = Some(Bounds(Punctuated::new()));
+                            } else {
+                                self.bound = Some(lit.parse().unwrap_or_else(|e| {
+                                    panic!("error parsing bounds: {:?} ({})", lit, e)
+                                }));
+                            }
+
+                            return;
+                        }
+                    }
+
+                    panic!(concat!(
+                        "The #[zeroize(bound)] attribute expects a name-value syntax with a string literal value.",
+                        "E.g. #[zeroize(bound = \"T: MyTrait\")]."
+                    ))
```

Comment:
This much error handling really makes me wish we were using `proc-macro-error`, although I haven't figured out how to make it play nicely with `synstructure`.

Reply:
I've tended to not use a lot of libraries when working with proc-macros exactly because of interoperability. I was close to suggesting to drop synstrucutre completely in favor of just using plain syn to skip parsing the user input and let rustc just say what's wrong with the where bounds.

I can look into it if you deem it necessary.

Reply:
Yeah, I recently dropped `synstructure` from another proc macro crate, although `zeroize_derive` is better leveraging it.

`proc-macro-error` though I think is definitely worth it, since it allows emitting errors with diagnostic information that includes the original source spans, which makes error messages generated from proc macros substantially better.

Reply:
I did it without proc-macro-error by just using [`to_compile_error`](https://docs.rs/syn/1.0.81/syn/parse/struct.Error.html#method.to_compile_error) until now. But our current strategy using panic isn't ideal, true enough.

Reply:
I'm assuming you meant elsewhere outside the scope of this PR?

With `proc-macro-error` I've managed to make sure that practically every error case is tied to the input code that caused it.

Reply:
> I'm assuming you meant elsewhere outside the scope of this PR?

Sry, lack of context indeed 😄.
> With `proc-macro-error` I've managed to make sure that practically every error case is tied to the input code that caused it.

That sounds great! I will take a look at it then.